File: core/tests/test_diode_loss_thermal.cpp
// Realistic diode loss + thermal — Catch2 tests.
//
// Validates Phase 1 of the inverter-bridge-losses work:
//   * IdealDiode now exposes V_F0 + R_d for a realistic forward-characteristic
//     linear fit, matching the powerStage YAML catalog convention.
//   * The runtime integrates V·I·dt per accepted timestep into a per-device
//     loss accumulator (average_power, peak_power, total_energy).
//   * A simple 1-stage R_th_ja thermal model maps P_avg → T_j.
//   * Circuit::add_bridge_rectifier helper wires four diodes as a full-wave
//     bridge so per-diode P_avg / T_j can be read back after the transient.
//
// Reference numbers come from a hand-calculated DC operating point — chosen
// because the bridge-rectifier waveform is highly nonlinear and hard to
// validate analytically without a current-controlled load.

#include <catch2/catch_test_macros.hpp>
#include <catch2/catch_approx.hpp>

#include "pulsim/v1/simulation.hpp"

#include <cmath>

using namespace pulsim::v1;
using Catch::Approx;

namespace {

SimulationOptions make_opts(Real tstop, Real dt) {
    SimulationOptions opts;
    opts.tstart = 0.0;
    opts.tstop = tstop;
    opts.dt = dt;
    opts.dt_min = 1e-9;
    opts.dt_max = dt;
    opts.adaptive_timestep = false;
    opts.enable_bdf_order_control = false;
    return opts;
}

}  // namespace

TEST_CASE("Diode V_F0 stamp: open-circuit operating point ≈ V_F0 across the diode",
          "[diode_loss][regression]") {
    // A diode in series with a small DC source: with V_source = 0.7 V exactly
    // at the GBU2506 MCC V_F0 corner, the conducted current should be near
    // zero (the Norton-shifted current source cancels the conductance term).
    Circuit circuit;
    const Index n_in = circuit.add_node("in");
    const Index gnd = Circuit::ground();

    IdealDiode::Params dp{};
    dp.V_F0    = 0.7;
    dp.R_d     = 0.01;
    dp.V_F0_tc = 0.0;             // no thermal feedback in this test
    dp.R_th_ja = 25.0;
    dp.T_amb   = 25.0;
    dp.g_on    = 100.0;  // 1/R_d

    circuit.add_voltage_source("Vsrc", n_in, gnd, 0.7);
    circuit.add_diode("D1", n_in, gnd, dp);

    auto opts = make_opts(2e-3, 50e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    Simulator sim(circuit, opts);
    const auto result = sim.run_transient();
    REQUIRE(result.success);

    // At V_source = V_F0, the steady-state diode current should be very
    // close to zero (within a few mA for the smoothed Behavioral path).
    const Real i_d = circuit.diode_last_current("D1");
    INFO("Diode current at V_source = V_F0 = 0.7 V: " << i_d << " A");
    CHECK(std::abs(i_d) < 0.5);  // within ~0.5 A of zero — Behavioral smoothing
                                  // accounts for the residual.
}

TEST_CASE("Diode V_F0 stamp: hard-driven diode tracks V_F0 + R_d · I",
          "[diode_loss][regression]") {
    // Drive 10 A through the diode and verify the voltage drop matches
    // V_F0 + R_d · I within 5 %. This is the linear-fit corner that
    // powerStage's catalog YAMLs target.
    constexpr Real I_drive = 10.0;
    constexpr Real R_series = 1.0;   // 1Ω series → V_source = 0.808 V + 10 V = ~10.8 V

    Circuit circuit;
    const Index n_in = circuit.add_node("in");
    const Index n_mid = circuit.add_node("mid");
    const Index gnd = Circuit::ground();

    IdealDiode::Params dp{};
    dp.V_F0    = 0.7;
    dp.R_d     = 0.01;
    dp.V_F0_tc = 0.0;             // no thermal feedback
    dp.g_on    = 100.0;
    dp.R_th_ja = 25.0;
    dp.T_amb   = 25.0;

    const Real v_src = R_series * I_drive + dp.V_F0 + dp.R_d * I_drive;
    circuit.add_voltage_source("Vsrc", n_in, gnd, v_src);
    circuit.add_resistor("Rs", n_in, n_mid, R_series);
    circuit.add_diode("D1", n_mid, gnd, dp);

    auto opts = make_opts(5e-3, 50e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    Simulator sim(circuit, opts);
    const auto result = sim.run_transient();
    REQUIRE(result.success);

    const Real v_diode = circuit.diode_last_voltage("D1");
    const Real i_diode = circuit.diode_last_current("D1");
    const Real expected_vd = dp.V_F0 + dp.R_d * i_diode;
    INFO("V_diode = " << v_diode << " V, I_diode = " << i_diode
         << " A, expected V_F0 + R_d·I = " << expected_vd << " V");
    CHECK(v_diode == Approx(expected_vd).epsilon(0.05));
    CHECK(i_diode == Approx(I_drive).epsilon(0.10));
}

TEST_CASE("Diode loss accumulator: average power = V_F · I_avg for steady DC",
          "[diode_loss][regression]") {
    // Series circuit: V_src → diode → R_load → GND. Steady-state current
    // I = (V_src − V_F0) / (R_d + R_load). The diode dissipates
    //   P = V_F · I = (V_F0 + R_d · I) · I.
    Circuit circuit;
    const Index n_anode = circuit.add_node("anode");
    const Index n_load  = circuit.add_node("load");
    const Index gnd = Circuit::ground();

    IdealDiode::Params dp{};
    dp.V_F0    = 0.7;
    dp.R_d     = 0.01;
    dp.V_F0_tc = 0.0;             // no thermal feedback
    dp.g_on    = 100.0;
    dp.R_th_ja = 25.0;
    dp.T_amb   = 25.0;

    const Real R_load = 1.0;
    const Real I_target = 5.0;
    const Real v_src = dp.V_F0 + (dp.R_d + R_load) * I_target;
    circuit.add_voltage_source("Vsrc", n_anode, gnd, v_src);
    circuit.add_diode("D1", n_anode, n_load, dp);
    circuit.add_resistor("Rload", n_load, gnd, R_load);

    auto opts = make_opts(20e-3, 100e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    Simulator sim(circuit, opts);
    const auto result = sim.run_transient();
    REQUIRE(result.success);

    const Real p_avg = circuit.diode_average_power("D1");
    const Real i_d = circuit.diode_last_current("D1");
    const Real v_d = circuit.diode_last_voltage("D1");
    const Real p_instant = v_d * i_d;
    const Real p_expected = (dp.V_F0 + dp.R_d * I_target) * I_target;
    INFO("Expected: I = " << I_target << " A, V_d = "
         << (dp.V_F0 + dp.R_d * I_target) << " V, P = " << p_expected
         << " W. Measured: I = " << i_d << ", V_d = " << v_d
         << ", P_avg = " << p_avg << ", P_instant = " << p_instant);
    CHECK(i_d == Approx(I_target).epsilon(0.10));
    CHECK(p_avg == Approx(p_expected).epsilon(0.15));
    CHECK(p_avg > 0.0);
}

TEST_CASE("Diode thermal: steady-state T_j = T_amb + P_avg · R_th_ja",
          "[diode_loss][regression]") {
    // Drive a steady current and verify the *derived* T_j matches the
    // analytical formula. `diode_junction_temperature` returns the value
    // assumed by the stamping (= T_amb at init for this pass), while
    // `diode_steady_state_junction_temperature` returns the value derived
    // from the accumulated P_avg.
    Circuit circuit;
    const Index n_anode = circuit.add_node("anode");
    const Index n_load  = circuit.add_node("load");
    const Index gnd = Circuit::ground();

    IdealDiode::Params dp{};
    dp.V_F0    = 0.7;
    dp.R_d     = 0.01;
    dp.g_on    = 100.0;
    dp.V_F0_tc = 0.0;            // disable T_j feedback for clean math
    dp.R_th_ja = 25.0;
    dp.T_amb   = 60.0;            // user's bench ambient

    const Real I_load = 5.0;
    const Real R_load = 1.0;
    const Real v_src = dp.V_F0 + dp.R_d * I_load + R_load * I_load;
    circuit.add_voltage_source("Vsrc", n_anode, gnd, v_src);
    circuit.add_diode("D1", n_anode, n_load, dp);
    circuit.add_resistor("Rload", n_load, gnd, R_load);

    auto opts = make_opts(50e-3, 100e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();
    Simulator sim(circuit, opts);
    const auto result = sim.run_transient();
    REQUIRE(result.success);

    const Real p_avg = circuit.diode_average_power("D1");
    const Real t_j_derived =
        circuit.diode_steady_state_junction_temperature("D1");
    const Real t_j_expected = dp.T_amb + p_avg * dp.R_th_ja;
    INFO("P_avg = " << p_avg << " W, T_amb = " << dp.T_amb
         << " °C, R_th_ja = " << dp.R_th_ja << " K/W → T_j_expected = "
         << t_j_expected << " °C; derived T_j = " << t_j_derived << " °C");
    CHECK(t_j_derived == Approx(t_j_expected).epsilon(0.001));
    CHECK(t_j_derived > dp.T_amb);   // self-heating actually happened
    // The stamping's T_j_ stays at the initial value because we don't
    // feed back automatically:
    CHECK(circuit.diode_junction_temperature("D1") == Approx(dp.T_amb));
}

TEST_CASE("Diode electrothermal: fixed-point iteration converges to "
          "self-consistent (T_j, P_avg)",
          "[diode_loss][regression]") {
    // The user's powerStage workflow runs a fixed-point iteration:
    // simulate → read T_j → push T_j back into the device → re-simulate.
    // After 2–3 passes (V_F0 keeps dropping with temperature so P_avg
    // drops too), the iteration converges.
    Circuit circuit;
    const Index n_anode = circuit.add_node("anode");
    const Index n_load  = circuit.add_node("load");
    const Index gnd = Circuit::ground();

    IdealDiode::Params dp{};
    dp.V_F0    = 0.7;
    dp.R_d     = 0.01;
    dp.g_on    = 100.0;
    dp.V_F0_tc = -2e-3;          // realistic silicon T_C
    dp.R_th_ja = 25.0;
    dp.T_amb   = 60.0;

    const Real I_load = 5.0;
    const Real R_load = 1.0;
    const Real v_src = dp.V_F0 + dp.R_d * I_load + R_load * I_load;
    circuit.add_voltage_source("Vsrc", n_anode, gnd, v_src);
    circuit.add_diode("D1", n_anode, n_load, dp);
    circuit.add_resistor("Rload", n_load, gnd, R_load);

    auto opts = make_opts(20e-3, 100e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();

    Real t_j_prev = dp.T_amb;
    Real t_j_curr = dp.T_amb;
    Real p_avg = 0.0;
    int passes = 0;
    constexpr int max_passes = 5;
    for (passes = 0; passes < max_passes; ++passes) {
        Simulator sim(circuit, opts);
        const auto result = sim.run_transient();
        REQUIRE(result.success);

        t_j_prev = t_j_curr;
        t_j_curr = circuit.diode_steady_state_junction_temperature("D1");
        p_avg = circuit.diode_average_power("D1");

        // Push the derived T_j back into the stamping for the next pass.
        circuit.set_diode_T_j("D1", t_j_curr);
        circuit.reset_diode_loss("D1");

        if (std::abs(t_j_curr - t_j_prev) < 0.5) break;  // < 0.5 K converged
    }
    INFO("Converged after " << passes
         << " passes: T_j = " << t_j_curr << " °C, P_avg = " << p_avg << " W");
    CHECK(passes <= max_passes - 1);  // some convergence happened
    CHECK(t_j_curr > dp.T_amb);
    // Self-consistency: T_j should equal T_amb + P_avg · R_th_ja.
    CHECK(t_j_curr == Approx(dp.T_amb + p_avg * dp.R_th_ja).epsilon(0.05));
}

TEST_CASE("Bridge rectifier helper: composes 4 diodes with shared params",
          "[diode_loss][bridge][regression]") {
    Circuit circuit;
    const Index ac_a = circuit.add_node("AC_A");
    const Index ac_b = circuit.add_node("AC_B");
    const Index dc_p = circuit.add_node("DC+");
    const Index dc_n = circuit.add_node("DC-");

    const std::size_t dev_before = circuit.devices().size();
    circuit.add_bridge_rectifier("BR", ac_a, ac_b, dc_p, dc_n, 0.7, 0.01);
    const std::size_t dev_after = circuit.devices().size();

    // 4 diodes added.
    CHECK(dev_after - dev_before == 4);

    // Each diode is reachable by its `<name>__D{1..4}` mangled name and
    // exposes the loss accessors.
    for (const char* suffix : {"__D1", "__D2", "__D3", "__D4"}) {
        const std::string n = std::string("BR") + suffix;
        // Before the simulation runs, all accessors return 0 (or T_amb).
        CHECK(circuit.diode_average_power(n) == Approx(0.0));
        CHECK(circuit.diode_total_energy(n) == Approx(0.0));
        CHECK(circuit.diode_peak_power(n) == Approx(0.0));
    }
}

TEST_CASE("Bridge rectifier: cap-input AC → DC produces non-zero per-diode P/T_j",
          "[diode_loss][bridge][regression]") {
    // Sanity-only end-to-end smoke test: drive the bridge with a 50 Hz sine
    // into an RC load, run for ~80 ms, and verify each diode sees some
    // conduction (the alternating pair pattern matches the topology). Exact
    // magnitudes match the cap-charging current pulse shape, which is hard
    // to validate analytically; that's left to the Python validation
    // workflow.
    Circuit circuit;
    const Index ac_a = circuit.add_node("AC_A");
    const Index dc_p = circuit.add_node("DC+");
    const Index dc_n = circuit.add_node("DC-");
    const Index gnd  = Circuit::ground();

    constexpr Real V_AC_peak = 100.0 * 1.4142135623730951;  // 100 V_rms

    SineParams sine{};
    sine.amplitude = V_AC_peak;
    sine.frequency = 50.0;
    sine.offset = 0.0;
    sine.phase = 0.0;
    circuit.add_sine_voltage_source("Vac", ac_a, gnd, sine);

    IdealDiode::Params dp{};
    dp.V_F0    = 0.7;
    dp.R_d     = 0.01;
    dp.g_on    = 100.0;
    dp.R_th_ja = 15.0;
    dp.T_amb   = 60.0;

    circuit.add_bridge_rectifier("BR", ac_a, gnd, dc_p, dc_n, dp);
    // RC output filter sized for ~100 W load: R_load = V²/P ≈ 140V²/100W ≈ 200Ω
    circuit.add_capacitor("Cbus", dc_p, dc_n, 220e-6);
    circuit.add_resistor("Rload", dc_p, dc_n, 200.0);
    // Anchor DC- to ground via small R so the solver has a closed loop.
    circuit.add_resistor("Rgnd", dc_n, gnd, 1e6);

    auto opts = make_opts(0.08, 50e-6);
    opts.newton_options.num_nodes = circuit.num_nodes();
    opts.newton_options.num_branches = circuit.num_branches();

    Simulator sim(circuit, opts);
    const auto result = sim.run_transient();
    REQUIRE(result.success);

    // Each diode in the bridge must have seen *some* conduction current
    // (positive average power, positive accumulated energy).
    for (const char* suffix : {"__D1", "__D2", "__D3", "__D4"}) {
        const std::string n = std::string("BR") + suffix;
        const Real p_avg = circuit.diode_average_power(n);
        const Real t_j = circuit.diode_junction_temperature(n);
        INFO("Diode " << n << ": P_avg = " << p_avg << " W, T_j = "
             << t_j << " °C");
        CHECK(std::isfinite(p_avg));
        CHECK(std::isfinite(t_j));
        CHECK(t_j >= dp.T_amb);  // self-heating only ever raises T_j
    }
}

File: python/tests/test_diode_loss_thermal.py
"""Smoke tests for realistic diode loss + thermal (Phase 1, Pulsim 0.10.0a6).

Validates the pybind11 surface area for the V_F0 + R_d linear-fit diode
model, the loss accumulator (P_avg / peak_P / E_total), the simple
R_th_ja thermal binding, and the bridge-rectifier helper. Mirrors the
Catch2 suite in core/tests/test_diode_loss_thermal.cpp.
"""
from __future__ import annotations

import math

import pytest

import pulsim as ps


def _run_transient(circuit, tstop=20e-3, dt=100e-6):
    opts = ps.SimulationOptions()
    opts.tstart = 0.0
    opts.tstop = tstop
    opts.dt = dt
    opts.dt_min = 1e-9
    opts.dt_max = dt
    opts.adaptive_timestep = False
    opts.enable_bdf_order_control = False
    # simplify-and-harden-numerical-surface §11: Auto resolves to Ideal in
    # v0.11. The V_F0 + R_d loss/thermal contract only validates on the
    # Behavioral diode stamp; PWL Ideal uses a unit-conductance switch and
    # would produce spurious test values. Pin Behavioral.
    opts.switching_mode = ps.SwitchingMode.Behavioral
    opts.newton_options.num_nodes = circuit.num_nodes()
    opts.newton_options.num_branches = circuit.num_branches()
    return ps.Simulator(circuit, opts).run_transient()


class TestRealisticDiodeBindings:
    """Verify the pybind11 surface for the new diode loss/thermal model."""

    def test_realistic_diode_params_exposed(self):
        p = ps.RealisticDiodeParams()
        p.V_F0 = 0.497   # GBU2506 MCC tangent fit @ T_j=125 °C
        p.R_d = 0.0108
        p.V_F0_tc = -2e-3
        p.R_th_ja = 25.0
        p.T_amb = 60.0
        assert p.V_F0 == pytest.approx(0.497)
        assert p.R_d == pytest.approx(0.0108)
        assert p.V_F0_tc == pytest.approx(-2e-3)
        assert p.R_th_ja == pytest.approx(25.0)
        assert p.T_amb == pytest.approx(60.0)

    def test_circuit_diode_methods_present(self):
        c = ps.Circuit()
        for name in (
            "diode_average_power", "diode_peak_power", "diode_total_energy",
            "diode_junction_temperature",
            "diode_steady_state_junction_temperature",
            "diode_last_current", "diode_last_voltage",
            "diode_conduction_time", "set_diode_T_j", "reset_diode_loss",
            "add_bridge_rectifier",
        ):
            assert hasattr(c, name), f"Circuit missing method {name!r}"


class TestRealisticDiodeBehavior:
    """End-to-end verification through the pybind boundary."""

    def test_loss_matches_VF_times_I_for_dc_steady_state(self):
        c = ps.Circuit()
        n_a = c.add_node("anode")
        n_l = c.add_node("load")
        gnd = ps.Circuit.ground()

        p = ps.RealisticDiodeParams()
        p.V_F0 = 0.7
        p.R_d = 0.01
        p.V_F0_tc = 0.0          # disable thermal feedback for clean math
        p.g_on = 100.0
        p.R_th_ja = 25.0
        p.T_amb = 25.0

        I_target = 5.0
        R_load = 1.0
        v_src = p.V_F0 + (p.R_d + R_load) * I_target

        c.add_voltage_source("Vsrc", n_a, gnd, v_src)
        c.add_diode("D1", n_a, n_l, p)
        c.add_resistor("Rload", n_l, gnd, R_load)

        r = _run_transient(c)
        assert r.success

        i_d = c.diode_last_current("D1")
        v_d = c.diode_last_voltage("D1")
        p_avg = c.diode_average_power("D1")
        p_expected = (p.V_F0 + p.R_d * I_target) * I_target
        assert i_d == pytest.approx(I_target, rel=0.10)
        assert p_avg == pytest.approx(p_expected, rel=0.15)
        # Steady-state T_j derivation is consistent with the formula.
        t_j = c.diode_steady_state_junction_temperature("D1")
        assert t_j == pytest.approx(p.T_amb + p_avg * p.R_th_ja, rel=0.01)

    def test_bridge_rectifier_composes_4_diodes(self):
        c = ps.Circuit()
        ac_a = c.add_node("ac_a")
        ac_b = c.add_node("ac_b")
        dc_p = c.add_node("dc_p")
        dc_n = c.add_node("dc_n")
        c.add_bridge_rectifier("BR", ac_a, ac_b, dc_p, dc_n, 0.7, 0.01)
        # 4 diodes appended.
        assert c.num_devices() == 4
        # Each diode is reachable by its mangled name.
        for suffix in ("__D1", "__D2", "__D3", "__D4"):
            n = "BR" + suffix
            assert c.diode_average_power(n) == pytest.approx(0.0)

    def test_bridge_rectifier_full_chain_produces_per_diode_telemetry(self):
        """AC source → 4-diode bridge → RC filter. Validates per-diode P_avg
        and T_j accessors return finite numbers after a short transient."""
        c = ps.Circuit()
        ac_a = c.add_node("ac_a")
        dc_p = c.add_node("dc_p")
        dc_n = c.add_node("dc_n")
        gnd = ps.Circuit.ground()

        # 100 V_rms, 50 Hz AC source between ac_a and gnd.
        sine = ps.SineParams()
        sine.amplitude = 100.0 * math.sqrt(2.0)
        sine.frequency = 50.0
        sine.offset = 0.0
        sine.phase = 0.0
        c.add_sine_voltage_source("Vac", ac_a, gnd, sine)

        # GBU2506 MCC tangent fit at T_j=125 °C.
        p = ps.RealisticDiodeParams()
        p.V_F0 = 0.497
        p.R_d = 0.0108
        p.V_F0_tc = 0.0
        p.R_th_ja = 15.0
        p.T_amb = 60.0
        p.g_on = 1.0 / p.R_d

        c.add_bridge_rectifier("BR", ac_a, gnd, dc_p, dc_n, p)
        c.add_capacitor("Cbus", dc_p, dc_n, 220e-6)
        c.add_resistor("Rload", dc_p, dc_n, 200.0)
        c.add_resistor("Rgnd", dc_n, gnd, 1e6)

        r = _run_transient(c, tstop=0.04, dt=50e-6)
        assert r.success

        for suffix in ("__D1", "__D2", "__D3", "__D4"):
            n = "BR" + suffix
            p_avg = c.diode_average_power(n)
            t_j = c.diode_steady_state_junction_temperature(n)
            assert math.isfinite(p_avg), f"{n}: P_avg not finite"
            assert math.isfinite(t_j), f"{n}: T_j not finite"
            assert t_j >= p.T_amb       # only self-heating raises T_j

    def test_fixed_point_thermal_iteration_converges(self):
        """Mimics the powerStage workflow: simulate → read T_j → push T_j
        back via set_diode_T_j → re-simulate. Should converge in ≤5 passes
        for a moderate-current diode."""
        c = ps.Circuit()
        n_a = c.add_node("anode")
        n_l = c.add_node("load")
        gnd = ps.Circuit.ground()

        p = ps.RealisticDiodeParams()
        p.V_F0 = 0.7
        p.R_d = 0.01
        p.V_F0_tc = -2e-3
        p.g_on = 100.0
        p.R_th_ja = 25.0
        p.T_amb = 60.0

        I_target = 5.0
        R_load = 1.0
        v_src = p.V_F0 + (p.R_d + R_load) * I_target
        c.add_voltage_source("Vsrc", n_a, gnd, v_src)
        c.add_diode("D1", n_a, n_l, p)
        c.add_resistor("Rload", n_l, gnd, R_load)

        t_j_prev = p.T_amb
        t_j_curr = p.T_amb
        p_avg = 0.0
        passes = 0
        for passes in range(6):
            r = _run_transient(c, tstop=20e-3, dt=100e-6)
            assert r.success
            t_j_prev = t_j_curr
            # Read both *before* resetting the accumulator so the
            # self-consistency assertion below holds.
            t_j_curr = c.diode_steady_state_junction_temperature("D1")
            p_avg = c.diode_average_power("D1")
            c.set_diode_T_j("D1", t_j_curr)
            c.reset_diode_loss("D1")
            if abs(t_j_curr - t_j_prev) < 0.5:
                break

        assert passes <= 5
        assert t_j_curr > p.T_amb
        # Self-consistency: T_j = T_amb + P_avg · R_th_ja within 5 %.
        assert t_j_curr == pytest.approx(
            p.T_amb + p_avg * p.R_th_ja, rel=0.05
        )
